libraries/primitive_consul_lock.rb

```diff
@@ -172,12 +171,60 @@ def to_json
       @h.to_json
     end
 
-    def concurrency
-      @h['concurrency']
-    end
-
     def holders
       @h['holders']
     end
   end
+
+  class SemaphoreNode < Semaphore
+    def self.get_or_create(path)
+      super(path)
+      SemaphoreNode.new(path, @desired_lock)
+    end
+
+    def enter(name, concurrency, options)
```

Comment:
there is a lot of code duplication between this class and SemaphoreRack. We should design the inheritance to increase code reuse.